include/boost/numeric/ublas/tensor/tensor_core.hpp

```diff
@@ -0,0 +1,736 @@
+     *  @param is zero-based indices where 0 <= is[r] < this->size(r) where  0 < r < this->rank()
+     */
+    template<class ... size_types>
+    [[nodiscard]] inline
+    constexpr reference at (size_type i, size_types ... is) {
+        if constexpr (sizeof...(is) == 0)
+            return this->data_[i];
+        else{
+            auto temp = detail::access<0ul>(size_type(0),this->strides_,i,std::forward<size_types>(is)...);
+            return this->data_[temp];
+        }
+    }
+
+    /** @brief Element access using a single index.
+     *
+     *
+     *  @code A(i) = a; @endcode
+     *
+     *  @param i zero-based index where 0 <= i < this->size()
+     */
+    [[nodiscard]] inline
+    constexpr const_reference operator()(size_type i) const {
+        return this->data_[i];
+    }
+
+
+    /** @brief Element access using a single index.
+     *
+     *  @code A(i) = a; @endcode
+     *
+     *  @param i zero-based index where 0 <= i < this->size()
+     */
+    [[nodiscard]] inline
+    constexpr reference operator()(size_type i){
+        return this->data_[i];
+    }
+
+    /** @brief Generates a tensor_core index for tensor_core contraction
+     *
+     *
+     *  @code auto Ai = A(_i,_j,k); @endcode
+     *
+     *  @param i placeholder
+     *  @param is zero-based indices where 0 <= is[r] < this->size(r) where  0 < r < this->rank()
+     */
+    template<std::size_t I, class ... index_types>
+    [[nodiscard]] inline
+    constexpr decltype(auto) operator() (index::index_type<I> p, index_types ... ps) const
+    {
+        constexpr auto N = sizeof...(ps)+1;
+        if( N != this->rank() )
+            throw std::runtime_error("Error in boost::numeric::ublas::operator(index::index_type<I>,index_types&&): "
+                "size of provided index_types does not match with the rank."
+            );
+
+        return std::make_pair( std::cref(*this),  std::make_tuple( p, std::forward<index_types>(ps)... ) );
+    }
+
+
+    /** @brief Reshapes the basic_tensor
+     *
+     *
+     * (1) @code A.reshape(extents{m,n,o});     @endcode or
+     * (2) @code A.reshape(extents{m,n,o},4);   @endcode
+     *
+     * If the size of this smaller than the specified extents than
+     * default constructed (1) or specified (2) value is appended.
+     *
+     * @note rank of the basic_tensor might also change.
+     *
+     * @param e extents with which the basic_tensor is reshaped.
+     * @param v value which is appended if the basic_tensor is enlarged.
+     */
+    inline
+    void reshape (extents_type const& e, value_type v = value_type{})
+    {
+        static_assert(is_dynamic_v<extents_type>,
+            "Error in boost::numeric::ublas::basic_tensor: static extents cannot be reshaped");
+        this->extents_ = e;
+        this->strides_ = strides_type(this->extents_);
+
+        auto p = product(extents_);
+        if(p != this->size())
+            this->data_.resize (p, v);
+    }
```

Comment:
should be a free standing function.

Reply:
This will be a free function in the next pull request because I didn't want to mix this pull request with freeing the function.

Reply:
👍